Broadcast Grain Volume and Centroid Info

permcody opened the pull request:
This PR makes sure that volume and centroid information is available on all ranks.

This also fixes a bug with GrainTracker occurring due to `_max_curr_grain_id` not being parallel consistent as well.

@dschwen the real work here is in the second patch. The first patch is mostly reorganization (plus the small bug fix) and unfortunately some large scale re-indentation which causes the huge diff.

permcody committed:
Reorganize a few methods to clean up finalize
Also make most methods collective instead of having a mismatch

permcody committed:
Broadcast volume and centroid info

closes #7703

permcody commented:
@SudiptaBiswas - This patch should get you up and running.

SudiptaBiswas commented:
This definitely improved my tests. I only see a weird case of volume being zero on a processor for one of the grains, when more than two processors are used. See below, 
```
processor_id    grain_id    volume     centroid (x,y)
    0              1         375       (30, 18.4167)
    1              1         375       (30, 18.4167)
    2              1         0         (0, 0)
``` 
For grain 0 all the ranks give same values.

SudiptaBiswas commented:
Okay, the error was because of a bug in my code. Fixed it. The tests are doing fine now. I will wait for this to get merged before pushing my changes.

permcody committed:
Make sure that halos widths are correct for EBSD

The EBSD halos have a depth that is one less than
regular sims. This is due to the expandEBSDGrains
call messing with the grain's local ids. This
commit makes sure that the halo width is adjusted for
this case.

permcody commented on modules/phase_field/src/postprocessors/GrainTracker.C:
```diff
@@ -1369,7 +1378,7 @@ GrainTracker::updateFieldInfo()
         const Elem * elem = mesh.elem(entity);
         std::vector<Point> centroid(1, elem->centroid());
-        if (_ebsd_reader)
+        if (_ebsd_reader && _first_time)
```
@tonkmr, @frombs - I was missing this little guy so the `UNIQUE_REGION` and `VARIABLE_COLORING` was unstable for EBSD runs. This PR should fix that. Additionally the halo level was only 1 which isn't thick enough to avoid coalescence so that's two problems.

Now the GrainTracker seems to be working a lot better with EBSD data with this patch. However there's now a new issue that's occuring on the example problem. There's a "split" grain on the left hand side of the boundary that's just close enough that the grain tracker sees it as a single grain in the initial dataset. On the second time step the two pieces have shrunk enough so that the GrainTracker thinks a new grain is forming there. 

![ebsd_data](https://cloud.githubusercontent.com/assets/3036960/18369425/e51b2ba2-75e2-11e6-892b-112f52b65482.png)
 
That's just fine from the GrainTracker point of view but it's raising havoc with the RGB object. It "should" be possible to detect this case and avoid creating a new grain in this case but I'll need to work on that. 

@frombs - Grab this branch and try your dataset again. Sorry for the constant changes but these are tough problems!

permcody replied:
I've looked at this problem quite a bit and I don't think it's easily solvable without a lot more work. I suggest for now that we move on and come back to this one. I'll open a ticket to deal with splitting grains. @frombs - I don't think this will prevent you from running your simulations.

permcody commented:
I'll take another look at the reconstruction tests.

Yes I'm aware that the EBSD example crashes. That' because the RGB object
is trying to request a color for the new grain it sees. If you simply turn
off that object, the simulation runs just fine. The issues is that I need
to figure out a way to handle one of these "splitting grains" (not to be
confused with an already split grain). I think I have a plan, I just need
to find some time to get it implemented. You should hopefully be able to
run your case because it doesn't have that problem.

On Thu, Sep 8, 2016 at 10:50 PM frombs <notifications@github.com> wrote:

> @permcody <https://github.com/permcody> - If you reduce "dt" from 10.0 to
> 3.0 in the TimeStepper block, the split grain will not go away after a
> single time step. A value of 10 results in rapid evolution of the grains.
> The simulation crashes after 6 time steps with the following error:
> *** ERROR ***
> Error! Index out of range in EBSDReader::indexFromIndex()
> [9]
> /Users/frombs/projects/moose/modules/phase_field/src/userobjects/EBSDReader.C,
> line 275, compiled Sep 8 2016 at 18:00:29
> application called MPI_Abort(MPI_COMM_WORLD, 1) - process 9
> ------------------------------
>
> I also had 2 GrainTracker and 3 reconstruction tests fail in the
> phase_field module after compiling.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/7706#issuecomment-245818924>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AC5XIH5HiO_Y01H2F7hHOCv_fooA7dp1ks5qoOWmgaJpZM4J4Zwo>
> .
>

frombs commented:
@permcody - By removing the RGB block the example runs as you suggested.

I also ran my test case and there is no change. The initial condition coming from GrainTracker looks correct if viewing the "gr_op" variable but the other maps (var_indices, unique_grains, ebsd_numbers, RGB_) do not match-up with the gr_op map.

permcody committed:
Fix bug with setting old/older values in solution

Don't try to read/write values to old or older values
when running a steady test. While this is highly unusual
with an evolution problem. There are test problems that
only do reconstruction.

SudiptaBiswas commented:
Looks like this PR might take a while to get merged. So, I pushed my updated objects along with 3d062f4 & 8b40d0b to my PR to see how the overall tests go. I hope that's okay.

permcody commented:
@frombs - One more bug fix for the crashing EBSD tests (turns out you can't read/write old/older values for "Steady" solves).

OK - there are so many complications at play here so let me see if I can explain some of things that you are seeing. First off, yes "gr_op" is the first field to look at. It uses the EBSD elemental data directly combined with the `ReconVarIC` op assignment so it really should be solid. 

Next, The var_indices field. This field will match up to the gr_op field if and only if there aren't any remaps on the initial condition. It turns out that almost all of these tests have remaps on the initial condition so those fields will almost never match. The unique_grains field is really the "grain index" field or in @dschwen's definition it's either the "global_id or local_id" value depending on whether or not you are "considering phase" - this field should never flip back and forth between values but it almost never corresponds to EBSD numbers or any other field.

The ebsd_numbers is the "feature_id", not to be confused with the grain index field.

I feel like we really need a better test suite here. We have 3 different EBSD sets. One of them is a really simple single phase diagram with only 9 or 10 grains. It's trivial (which makes a good test), it even shows that grain index and ebsd numbers can be different. I think I'd like a 2-phase data set that's nearly as simple. We could run it with and without the phase options to `ReconVarIC` and `GrainTracker` (btw - these parameters should always match between these objects). For these "reconstruction" only type tests the GrainTracker should be there but the "remap_grains" option should really be turned off. This will help verify that the "gr_op" field and "var_idx" field are identical. Finally we can build some new evolution examples.

One more thing I've noticed. If you use the phase option, then you'll see slightly different fields between "gr_op" and "var_idx". The reason is that in the place where you have a different phase, the active phase can bleed over a little into the place where the inactive phase is at. It's a little confusing to look at but that's the reason for the variance.

@dschwen - Have I said anything that doesn't make sense? If you agree with these changes lets get this branch merged as soon as it finishes. I'm not seeing anything on any of these tests now that raise any concerns as to the way the grain tracker is functioning.

NOTE: The EBSD example is still broken for right now. I still need to handle the "splitting grain" case. I'll try to get that in soon.

As you've seen these directories are currently a mess.

permcody commented:
Sweet! Finally, all tests passed

permcody committed:
Replace a local map with vector

The new_grain_idx_to_existing_grain_idx is densely
populated. Switching it to a vector instead.

permcody committed:
Fixing error with accidental insertion EBSDReader

The semi-local node to element map doesn't contain all
entries needed for producing the necessary information
on multiple processors. Instead of adding blank entries
just skip the loop
